contrastive_vi/module/contrastive_vi.py

```diff
@@ -578,60 +551,37 @@ def loss(
+        kl_divergence_s = target_losses["kl_s"]
+        kl_divergence_l = background_losses["kl_library"] + target_losses["kl_library"]
+
+        wasserstein_loss = (
+            torch.norm(inference_outputs["background"]["qs_m"], dim=-1)**2
+            + torch.norm(torch.sqrt(inference_outputs["background"]["qs_v"]), dim=-1)**2
         )
 
-        if self.disentangle:
-            z_tar = inference_outputs["target"]["qz_m"]
-            s_tar = inference_outputs["target"]["qs_m"]
-
-            # If more than one sample, the outputs have dimension
-            # (n_samples, batch_size, n_latent). Otherwise, the outputs have dimension
-            # (batch_size, n_latent). We want to make sure that the first dimension
-            # corresponds to the batch size for total correlation estimation.
-            if len(z_tar.shape) == 3:
-                z_tar = z_tar.permute(1, 0, 2)
-                s_tar = s_tar.permute(1, 0, 2)
-            z1, z2 = torch.chunk(z_tar, 2)
-            s1, s2 = torch.chunk(s_tar, 2)
-
-            # Make sure all tensors have same number of batch samples. This is
-            # necessary e.g. if we have an odd batch size at the end of an epoch.
-            size = min(len(z1), len(z2))
-            z1, z2, s1, s2 = z1[:size], z2[:size], s1[:size], s2[:size]
-
-            q = torch.cat([torch.cat([z1, s1], dim=-1), torch.cat([z2, s2], dim=-1)])
-            q_bar = torch.cat(
-                [torch.cat([z1, s2], dim=-1), torch.cat([z2, s1], dim=-1)]
-            )
-            q_bar_score = F.sigmoid(self.discriminator(q_bar))
-            q_score = F.sigmoid(self.discriminator(q))
-            tc_loss = torch.log(q_score / (1 - q_score))
-            discriminator_loss = -torch.log(q_score) - torch.log(1 - q_bar_score)
-            loss += torch.sum(tc_loss) + torch.sum(discriminator_loss)
-
-        if self.use_mmd:
-            z_tar = inference_outputs["target"]["qz_m"]
-            z_background = inference_outputs["background"]["qz_m"]
-            mmd = self.mmd_loss(z_tar, z_background)
-            loss += self.mmd_weight * torch.sum(mmd)
+        kl_local_for_warmup = kl_divergence_z + kl_divergence_s
+        kl_local_no_warmup = kl_divergence_l
+
+        weighted_kl_local = kl_weight * (self.wasserstein_penalty*wasserstein_loss
+            + kl_local_for_warmup) + kl_local_no_warmup
+
+        loss = torch.mean(reconst_loss + weighted_kl_local)
 
         kl_local = dict(
-            kl_z=kl_z,
-            kl_s=kl_s,
-            kl_library=kl_library,
+            kl_divergence_l=kl_divergence_l,
+            kl_divergence_z=kl_divergence_z,
+            kl_divergence_s=kl_divergence_s
         )
         kl_global = torch.tensor(0.0)
-        return LossRecorder(loss, recon_loss, kl_local, kl_global)
+        return LossRecorder(
+            loss,
+            reconst_loss,
+            kl_local,
+            kl_global,
+            wasserstein_loss=torch.sum(wasserstein_loss)
```

Comment:
AFAICT, the `LossRecorder` internally sums any terms that aren't already a scalar before logging them. I.e., `loss` remains untouched (since we took the mean previously), but `recont_loss`, `kl_local`, and `kl_global`, are summed before being logged.

`LossRecorder` requires that any additional arguments for logging have already been reduced to scalars, so for consistency with the other individual components of the loss I took the sum of `wasserstein_loss`.

Reply:
I added a new comment to the code clarifying this.